graphene_federation/service.py

```diff
@@ -91,30 +102,115 @@ def get_sdl(schema: Schema) -> str:
+            _schema_import.append('"@override"')
+        if provides_parent_types or provides_fields:
+            _schema_import.append('"@provides"')
+        if required_fields:
+            _schema_import.append('"@requires"')
+        if inaccessible_types or inaccessible_fields:
+            _schema_import.append('"@inaccessible"')
+        if shareable_types or shareable_fields:
+            _schema_import.append('"@shareable"')
+        if tagged_fields:
+            _schema_import.append('"@tag"')
+        schema_import = ", ".join(_schema_import)
+        _schema = f'extend schema @link(url: "https://specs.apollo.dev/federation/v2.0", import: [{schema_import}])\n'
+
+    # Add fields directives (@external, @provides, @requires, @shareable, @inaccessible)
+    entities_ = (
+        set(provides_parent_types.values())
+        | set(extended_types.values())
+        | set(entities.values())
+        | set(required_fields.values())
+        | set(provides_fields.values())
+    )
 
-    # Add fields directives (@external, @provides, @requires)
-    for entity in set(provides_parent_types.values()) | set(extended_types.values()):
+    if schema.federation_version == 2:
+        entities_ = (
+            entities_
+            | set(shareable_types.values())
+            | set(inaccessible_types.values())
+            | set(inaccessible_fields.values())
+            | set(shareable_fields.values())
+            | set(tagged_fields.values())
+        )
+    for entity in entities_:
         string_schema = add_entity_fields_decorators(entity, schema, string_schema)
 
     # Prepend `extend` keyword to the type definition of extended types
+    # noinspection DuplicatedCode
     for entity_name, entity in extended_types.items():
-        type_def = re.compile(r"type %s ([^\{]*)" % entity_name)
-        repl_str = r"extend type %s \1" % entity_name
+        type_def = re.compile(rf"type {entity_name} ([^{{]*)")
+        repl_str = rf"extend type {entity_name} \1"
         string_schema = type_def.sub(repl_str, string_schema)
 
     # Add entity keys declarations
     get_field_name = type_attribute_to_field_name(schema)
     for entity_name, entity in entities.items():
-        type_def_re = r"(type %s [^\{]*)" % entity_name
-        type_annotation = " ".join(
-            [f'@key(fields: "{get_field_name(key)}")' for key in entity._keys]
-        )
-        repl_str = r"\1%s " % type_annotation
+        type_def_re = rf"(type {entity_name} [^\{{]*)" + " "
+        if hasattr(entity, "_resolvable") and not entity._resolvable:
+            type_annotation = (
+                (
+                    " ".join(
+                        [
+                            f'@key(fields: "{get_field_name(key)}"'
+                            for key in entity._keys
+                        ]
+                    )
+                )
+                + f", resolvable: {str(entity._resolvable).lower()})"
+                + " "
+            )
```

Comment:
I think we should do this only if federation 2 is enabled

Reply:
Yes. This is updated with a federation version check.